File: 백준/Gold/1238. 파티/파티.py
#n개의 마을과 학생, 각 마을에 한명씩 산다
#x번 마을 왕복(오고가는 시간 다름))
#m개의 단방향 도로 i번째 길을 지나는데 T(i)만큼의 시간 소비
#오고가는 각각의 최단시간 중 가장 오래 걸린 시간 출력
import heapq

n,m,x=map(int,input().split())
INF=int(1e9)
graph=[[] for _ in range(n+1)]

#간선 가중치가 다른 최단거리
for i in range(m):
    s,e,t=map(int,input().split())
    #받을때부터 (시간, 위치)순으로 값 삽입
    graph[s].append((t,e))

# print(graph)
# [[], [(2, 4), (3, 2), (4, 7)], [(1, 1), (3, 5)], [(1, 2), (4, 4)], [(2, 3)], [], [], []]

def dijkstra(start):
    q=[]
    distance=[INF for _ in range(n+1)] #1~n
    
    heapq.heappush(q, (0,start))
    distance[start]=0
  
    while q:
        time, now=heapq.heappop(q)
    
        for i in graph[now]:
              next_t, next=i
              if distance[next]> time+next_t:
                  distance[next]=time+next_t
                  heapq.heappush(q, (distance[next], next))
      
    return distance #배열로 반환

result=0
#각 학생들의 거리 계산을 위한 for문
for i in range(1,n+1):
    #dijkstra(i) : i번째 마을에서 시작해서 각 마을까지의 최단거리 배열
    #dijkstra(i)[x] : 그 중 x마을까지 가는데 걸린 시간
    go=dijkstra(i)[x] 
    back=dijkstra(x)[i] # x마을에서 i번째로 돌아오기까지의 최단시간
  
    result=max(result, go+back)

print(result)

File: 백준/Gold/1238. 파티/파티.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

class Town implements Comparable<Town>{
    int end;
    int time;

    public Town(int end, int time) {
        this.end = end;
        this.time = time;
    }

    @Override
    public int compareTo(Town o) {
        return time - o.time;
    }
}

public class Main {
    private static final int INF=Integer.MAX_VALUE;
    //n개 마을, 학생: x번째 마을까지 왕복 / 도로 m개
    static int n,m,x,s,e,t,result;
    static List<ArrayList<Town>> graph;
    static int[] distance;

    public static void main(String[] args) throws IOException {
        BufferedReader br= new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st=new StringTokenizer(br.readLine());

        n=Integer.parseInt(st.nextToken());
        m=Integer.parseInt(st.nextToken());
        x=Integer.parseInt(st.nextToken());

        graph=new ArrayList<>();
        for (int i = 0; i < n+1; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            st=new StringTokenizer(br.readLine());
            s=Integer.parseInt(st.nextToken());
            e=Integer.parseInt(st.nextToken());
            t=Integer.parseInt(st.nextToken());

            //생성자 순서에 맞게 넣어줄것
            graph.get(s).add(new Town(e,t));
        }

        result=0;
        for (int i = 1; i < n+1; i++) {
            int[] go=dijkstra(i);
            int[] back=dijkstra(x);
            //i에서 x까지, x에서 i까지
            result=Math.max(result, go[x]+back[i]);
        }
        System.out.println(result);

    }

    private static int[] dijkstra(int start){
        PriorityQueue<Town> q=new PriorityQueue<>();
        distance= new int[n+1];
        Arrays.fill(distance, INF);

        distance[start]=0;
        //생성자를 end, time순으로 해놓고 값을 반대로 삽입해서 오류 발생! 이 부분 주의할 것
        q.offer(new Town(start, 0));

        while (!q.isEmpty()){
            Town t1=q.poll();
            int now_t=t1.time;
            int now=t1.end;

            for (Town t2: graph.get(now)) {
                int next_t=t2.time;
                int next=t2.end;

                if (distance[next]> now_t+next_t){
                    distance[next]= now_t+next_t;

                    //end, time 순 지켜서 넣기
                    q.offer(new Town(next, distance[next]));
                }
            }
        }
        return distance;
    }
}